### PR TITLE
poc: git backend

### LINKED ISSUE #5330: Git SoT Phase #5: git-first backend
Phase <span>#</span>5 of the Git Source-of-Truth (Git SoT) stack.

Swap the draft / commit / publish backend from DB-native to **git-backed (supergit)** without changing the editing UX delivered in Phases <span>#</span>1–<span>#</span>4. The staging → commit → publish loop, repository-files spec API, and draft RPCs keep their names and behavior — only their implementations change. After this phase, committed specs files (`canvas.yaml`, `console.yaml`) will live in git, version IDs become commit SHAs, and `workflow_versions` rows become materialized projections of git commits.

POC: #5153

This is a **breaking change** (`feat!`): version IDs change from UUIDs to commit SHAs and the DB-only write paths are removed. Ship as a **single PR** — do not split the backend swap across multiple merges to `main`, to avoid mixed DB/git write paths.

### Infrastructure (internal git only — not user-facing)

* Bump supergit to **0.1.2** in `docker-compose.dev.yml` (currently 0.1.1) — see supergit#2. `GIT_STORAGE_SUPERGIT_BASE_URL` already exists.
* Extend `pkg/git/provider.Provider` with branch operations: `ListBranches`, `CreateBranch`, `MergeBranch`, `DeleteBranch`. (`Head`, `ListFiles`, `GetFile`, `Commit` already exist.)
* Implement the new operations in `pkg/git/supergit`, `pkg/git/codestorage`, and `pkg/git/inmemory`.
* `START_REPOSITORY_MATERIALIZER=yes` wiring (worker startup in `cmd/server/main.go` + docker compose env).

### Database & migration

* Migration `workflow_versions`: keep UUID id, add `commit_sha` column, add materialization columns.
* New `repository_materialization_state` table.
* `workflow_staging`: add `base_head_sha`; backfill from existing `base_version_id`.
* Backfill: turn each existing draft branch (Phase <span>#</span>1) into a real git branch and seed `canvas.yaml` / `console.yaml` from the current version row.

### Materialization

* New `pkg/canvas/materialize/*` and `pkg/canvas/yaml/*` packages (git ⇄ `workflow_versions` projection).
* `CreateCanvas` seeds the git repo and materializes the live version.
* `RepositoryMaterializerWorker` reconciles DB state from git via the internal supergit API.
* Materialization in the DB needed only for the live versions of the app.

### Swap handlers to git (same RPCs as Phases <span>#</span>1–<span>#</span>4)

* `CreateDraftBranch` → create a git branch from `main` + sync the draft branch from git.
* `DeleteDraftBranch` → delete the git branch + reconcile draft deletions from git.
* `CommitCanvasStaging` → staging → `CommitCanvasRepositoryFiles` (git commit) → materialize → clear staging.
* `PublishCanvas` → merge the draft branch into `main` + sync live from git.
* `canvas.yaml` / `console.yaml` become git-backed virtual files instead of DB-only.

### Change management

* `pkg/canvas/changerequests/*`: diff draft SHA vs live; publish via git merge.

### Read paths

* Editor baseline: staging overlay overrides the **git tip** (instead of the draft version row).
* `ListCanvasRepositoryFiles` accepts `branch` / `ref`.
* Files tab reads arbitrary repo files from git for any branch.

### Cleanup

* Remove the internal `UpdateCanvasVersion`, `UpdateConsole`, `GetConsole` Go helpers retained for the DB-native commit path (public RPCs were already removed in Phase <span>#</span>2).
* Remove `ApplyCanvasVersionChangeset` and `CreateCanvasVersion` handlers.
* Remove the DB-only commit/publish code paths.

### Agent (prompt only — no new CLI)

* Report the resulting commit **SHA** in `:::draft-actions` after commit.
* Re-fetch after publish notifications.

## Out of Scope

Deferred to [Phase <span>#</span>6 (git platform)](https://github.com/superplanehq/superplane/issues/5331):

* Public `/git/{canvas-id}.git` proxy and external clone/push/pull.
* `GIT_STORAGE_SUPERGIT_GIT_URL` / `SUPERGIT_PUBLIC_URL` for external clients.
* `superplane clone` and repository-URL CLI commands.
* `docs/contributing/git-native-apps.md`.

### PR DESCRIPTION
**Pull request by mchalapuk**
## Summary

Migrates the canvas draft / commit / publish backend from DB-native storage to **git-backed storage**, treating git as the source of truth and `workflow_versions` as a materialized projection of git branches and commits. This is Phase <span>#</span>5 of the Git Source-of-Truth effort.

Closes #5330

### Git as source of truth
- Drafts are real **git draft branches**. Per-draft edits are held in `workflow_staged_files` (with `base_head_sha`) and materialized into version rows on commit.
- Change requests publish via **git merge into `main`**, then the live version is materialized from the resulting commit.
- `workflow_versions.id` stays a UUID and existing RPC names are preserved (`CreateCanvasVersion`, `PublishCanvasVersion`, etc.); drafts remain `workflow_versions` rows.

### Materialization (`pkg/canvas/materialize`)
- New seed / sync / backfill pipeline: seed repositories and `canvas.yaml` / `console.yaml`, sync live and draft branches into DB projections, reconcile deleted draft branches, and backfill existing pre-git-first canvases (draft rows → real git branches).
- New `repository_materializer` worker and a `repository_branch` event distributer; new `RepositoryBranchUpdated` message so the UI reacts to branch updates.

### Git provider
- Extended the provider interface and in-memory / codestorage / supergit implementations (branch create/merge/head, commit with expected-head, etc.).

### Change requests
- Consolidated CR diff / conflict / refresh logic into `pkg/canvas/changerequests`.

### Cleanup
- Removed the deprecated DB-native write paths now superseded by git-first: the `ApplyCanvasVersionChangeset` / `ValidateCanvasVersionChangeset` RPCs (proto, service methods, authorization entries) and the internal `UpdateCanvasVersion(WithUsage)` / `UpdateConsole` handlers, plus their tests.
- Relocated still-used helpers out of the deleted files and removed the now-dead client-side AI changeset apply path in the UI.

### Migration
- `20260612120000_git-first-canvas-versions`: adds git-first columns to `workflow_versions` / `workflow_staged_files` and a `repository_materialization_state` table.

## Test plan

- [x] create canvas → edit draft → commit → publish; works as before
- [x] Modify README.md → uncommited changes → commit → committedd to branch/draft; README.md in published version unchanged; blue dot next to Files tab → publish
- [x] Files Diff always compare against live version

### CODE REVIEW

File: pkg/authorization/service.go [hunk 1]
```diff
@@ -130,7 +136,16 @@ func (a *AuthService) newReadEnforcer(filters []gormadapter.Filter) (casbin.IEnf
...
+	// Each read enforcer gets its own model. casbin mutates the model during
+	// enforcer initialization and policy loading, so sharing a single model
+	// across concurrently created enforcers causes data races (and would
+	// clobber the shared enforcer's in-memory policies).
+	readModel, err := model.NewModelFromString(a.casbinModelText)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load casbin model: %w", err)
+	}
+
+	enforcer, err := casbin.NewEnforcer(readModel, adapter)
```
Comment (mchalapuk) [inline pkg/authorization/service.go:149]: Is this change in any way related to git backend?

File: pkg/authorization/interceptor.go [hunk 2]
```diff
@@ -259,18 +259,6 @@ func DefaultAuthorizationRules() map[string]AuthorizationRule {
...
-			Action:           "update_version",
-			DomainType:       models.DomainTypeOrganization,
-			ResourceResolver: canvasResourceResolver,
-		},
-		pbCanvases.Canvases_ValidateCanvasVersionChangeset_FullMethodName: {
-			Resource:         "canvases",
-			Action:           "read",
-			DomainType:       models.DomainTypeOrganization,
-			ResourceResolver: canvasResourceResolver,
-		},
```
Comment (mchalapuk) [inline pkg/authorization/interceptor.go:274]: Why did we remove those two?

File: pkg/canvas/materialize/seed_repository.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+)
+
+type SeedRepositoryInput struct {
+	Name                    string
+	Description             string
+	Nodes                   []models.Node
+	Edges                   []models.Edge
+	ChangeManagementEnabled bool
+	ChangeRequestApprovers  []models.CanvasChangeRequestApprover
+	Console                 *models.ConsoleYAML
```
Comment (mchalapuk) [inline pkg/canvas/materialize/seed_repository.go]: It's weird that we're passing typed data for canvas and yaml for console. Is there any reason for it?

File: pkg/canvas/materialize/sync_draft_branch.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+// GitBranchExists reports whether a branch ref exists in the git repository.
+func GitBranchExists(ctx context.Context, gitProvider git.Provider, repoID, branch string) bool {
+	_, err := gitProvider.Head(ctx, repoID, branch)
+	return err == nil
+}
+
+// SyncDraftBranchFromGit registers draft workflow_versions metadata when missing and
+// materializes the branch tip from git. Safe to call repeatedly.
+func SyncDraftBranchFromGit(
```
Comment (mchalapuk) [inline pkg/canvas/materialize/sync_draft_branch.go:27]: This should be the first function declared in this file.

File: pkg/canvas/materialize/sync_draft_branch.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+			return nil, createErr
+		}
+	}
+
+	mat := &DraftMaterializer{GitProvider: gitProvider, Registry: reg}
+	ownerID := draftVersion.OwnerID
+	version, matErr := mat.MaterializeDraft(ctx, tx, orgID, canvasID, branchName, headSHA, ownerID)
+	if matErr != nil {
+		return nil, matErr
+	}
```
Comment (mchalapuk) [inline pkg/canvas/materialize/sync_draft_branch.go:100]: Shouldn't materialization de done asynchronously (in a worker)?

File: pkg/grpc/actions/canvases/commit_canvas_repository_files.go [hunk 6]
```diff
@@ -36,6 +39,12 @@ func CommitCanvasRepositoryFiles(
 	webhookBaseURL string,
 	authService authorization.Authorization,
 ) (*pb.CommitCanvasRepositoryFilesResponse, error) {
+	_ = usageService
+	_ = encryptor
+	_ = webhookBaseURL
+	_ = authService
```
Comment (mchalapuk) [inline pkg/grpc/actions/canvases/commit_canvas_repository_files.go:45]: Does this mean the request will not be authorized?

File: pkg/grpc/actions/canvases/commit_canvas_repository_files.go [hunk 7]
```diff
@@ -151,8 +139,33 @@ func commitGitFileOperations(
...
+		// A direct commit re-materializes the draft from git, so any staged DB
+		// edits for this version are now stale and must be cleared.
+		if materialized != nil {
+			return models.DiscardWorkflowStagingInTransaction(tx, materialized.ID, nil)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to materialize draft: %v", err)
 	}
```
Comment (mchalapuk) [inline pkg/grpc/actions/canvases/commit_canvas_repository_files.go:174]: Why are we not doing this in the materialization worker?

File: pkg/grpc/actions/canvases/commit_canvas_staging.go [hunk 8]
```diff
@@ -192,114 +104,86 @@ func commitStagedGitFiles(
...
+	err = database.Conn().Transaction(func(tx *gorm.DB) error {
+		mat := &materialize.DraftMaterializer{GitProvider: gitProvider, Registry: registry}
+		_, matErr := mat.MaterializeDraft(ctx, tx, orgUUID, canvas.ID, draftBranch, commitSHA, version.OwnerID)
+		if matErr != nil {
+			return matErr
+		}
+		return models.DiscardWorkflowStagingInTransaction(tx, version.ID, nil)
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to materialize draft: %v", err)
```
Comment (mchalapuk) [inline pkg/grpc/actions/canvases/commit_canvas_staging.go]: Same. Why not in a worker?

File: pkg/grpc/actions/canvases/create_canvas.go [hunk 9]
```diff
@@ -129,116 +127,122 @@ func CreateCanvas(
...
-		})
-
-		if err != nil {
-			return err
 		}
-
-		return publisher.Publish(ctx)
+		_, matErr := mat.MaterializeFromGit(ctx, tx, organizationID, canvasID, models.CanvasGitBranchMain, commitSHA, materialize.ModeLive, &createdBy)
+		return matErr
 	})
```
Comment (mchalapuk) [inline pkg/grpc/actions/canvases/create_canvas.go]: Same. Why not in the worker?

File: pkg/grpc/actions/canvases/delete_canvas_version.go [hunk 10]
```diff
@@ -46,39 +50,67 @@ func DeleteCanvasVersion(
...
+	err = database.Conn().Transaction(func(tx *gorm.DB) error {
+		var reconcileErr error
+		removed, reconcileErr = materialize.ReconcileDraftBranchDeletionsFromGit(
+			ctx,
+			tx,
+			gitProvider,
+			canvasUUID,
+			materialize.ReconcileDraftBranchDeletionsOptions{BranchName: branchName},
+		)
+		return reconcileErr
```
Comment (mchalapuk) [inline pkg/grpc/actions/canvases/delete_canvas_version.go:106]: Same. Should be done in the worker.

File: pkg/grpc/actions/canvases/publish_canvas_version.go [hunk 11]
```diff
@@ -99,70 +100,108 @@ func publishDraftVersionInTransaction(
...
+			materialize.SyncLiveFromGitOptions{
+				HeadSHA:                   mergeSHA,
+				SkipChangeManagementCheck: true,
 			},
 		)
-		if err != nil {
-			log.Errorf("failed to publish canvas version: %v", err)
-			return err
+		if syncErr != nil {
+			return syncErr
```
Comment (mchalapuk) [inline pkg/grpc/actions/canvases/publish_canvas_version.go]: Worker...

File: pkg/grpc/actions/canvases/repository_spec_files.go [hunk 12]
```diff
@@ -174,14 +178,11 @@ func loadRepositorySpecVersionForRead(
...
 
-// ApplyRepositorySpecFileOperations parses canvas.yaml/console.yaml content into
-// the draft version row. It is the validated write path shared by the
-// staging-commit flow (CommitCanvasStaging) and the direct-commit flow
-// (CommitCanvasRepositoryFiles). When autoLayout is set it lays out canvas.yaml
-// during the write; when discardStaging is set it drops any staged edits for the
-// version in the same transaction as the version-row write.
+// ApplyRepositorySpecFileOperations commits canvas.yaml/console.yaml to the draft
+// branch in git, materializes the resulting commit, and optionally clears staging.
 func ApplyRepositorySpecFileOperations(
```
Comment (mchalapuk) [inline pkg/grpc/actions/canvases/repository_spec_files.go]: Now that all files (including spec) are stored in git and materialization happens in the worker, why do we need this function?

File: pkg/models/canvas_version.go [hunk 13]
```diff
@@ -92,6 +104,149 @@ func FindCanvasVersion(workflowID, versionID uuid.UUID) (*CanvasVersion, error)
 	return FindCanvasVersionInTransaction(database.Conn(), workflowID, versionID)
 }
 
+func FindVersionByCommitSHAInTransaction(tx *gorm.DB, workflowID uuid.UUID, commitSHA string) (*CanvasVersion, error) {
```
Comment (mchalapuk) [inline pkg/models/canvas_version.go:173]: This file is not compliant with file layout guidelines.

**Review by mchalapuk — COMMENTED** [review]
My impression is that it needs some more work.

File: pkg/authorization/service.go (hunk 1 above)
Comment (mchalapuk) [reply pkg/authorization/service.go:149]: Not directly — this is a latent Casbin concurrency fix. `newReadEnforcer` used to build enforcers from a shared model instance; Casbin mutates the model during initialization and policy loading, so concurrent permission checks raced (`concurrent map read and map write`) and could crash the process. Each read enforcer now gets its own model parsed from the cached definition text.

The git-first work did not introduce this bug, but the heavier concurrent request load in e2e/CI exposed it. The fix is orthogonal to git storage and could live in its own commit, but it was needed to get the branch green.

File: pkg/authorization/interceptor.go (hunk 2 above)
Comment (mchalapuk) [reply pkg/authorization/interceptor.go:274]: Those two entries were for `ApplyCanvasVersionChangeset` and `ValidateCanvasVersionChangeset`. Both RPCs were removed in the cleanup commit (`chore: remove deprecated DB-native canvas version write RPCs`) because the git-first staging/commit flow supersedes them — there are no remaining production callers, so the interceptor mappings became dead code.

File: pkg/grpc/actions/canvases/commit_canvas_repository_files.go (hunk 6 above)
Comment (mchalapuk) [reply pkg/grpc/actions/canvases/commit_canvas_repository_files.go:45]: No — the request is still authorized. The `_ = authService` (and the other `_ = …` lines) only silence parameters that are no longer used inside this handler after the refactor. Authorization for `CommitCanvasRepositoryFiles` is enforced at the gRPC interceptor layer (`pkg/authorization/interceptor.go`) before the handler runs, same as the other canvas RPCs.

File: pkg/grpc/actions/canvases/repository_spec_files.go (hunk 12 above)
Comment (mchalapuk) [reply pkg/grpc/actions/canvases/repository_spec_files.go]: Removed in 4b0554c41. It had no production callers — spec and arbitrary file commits already go through `CommitCanvasRepositoryFiles`. The canvas YAML edge-case tests now call that path directly via `commitCanvasRepositoryFilesForTest`.

File: pkg/canvas/materialize/seed_repository.go (hunk 3 above)
Comment (mchalapuk) [reply pkg/canvas/materialize/seed_repository.go]: Good catch — there was no intentional reason for the asymmetry; it was just how the two helpers evolved.

`SeedRepositoryInput` now takes both specs as composite objects:
- `Canvas *materialize.CanvasYAML` (new, mirrors the console shape)
- `Console *models.ConsoleYAML` (unchanged)

`CanvasYAMLFromVersion` and `BuildCanvasYAMLFromCanvas` handle the version-row and YAML-encoding paths. Addressed in e3d89d258.

File: pkg/canvas/materialize/sync_draft_branch.go (hunk 4 above)
Comment (mchalapuk) [reply pkg/canvas/materialize/sync_draft_branch.go:27]: Done in 27b57602c — `SyncDraftBranchFromGit` is now the first function after `SyncDraftBranchOptions`; the branch-name helpers (`OwnerFromDraftBranchName`, `NextDraftDisplayName`, `UniqueDraftBranchName`, `GitBranchExists`) follow it.

File: pkg/models/canvas_version.go (hunk 13 above)
Comment (mchalapuk) [reply pkg/models/canvas_version.go:173]: Done in 27b57602c — reordered per AGENTS.md: struct/constants → getters → conn wrappers (all grouped) → `…InTransaction` methods (all grouped) → private helpers at the end. No behavior change.

File: pkg/canvas/materialize/sync_draft_branch.go (hunk 5 above)
Comment (mchalapuk) [reply pkg/canvas/materialize/sync_draft_branch.go:100]: Yes — done in 36b857a69. Materialization is now worker-authoritative.

Canvas write handlers commit to git, register a pending `workflow_versions` row, and call `materialize.RequestBranchMaterialization`, which publishes a `repository_branch_updated` message for the `RepositoryMaterializerWorker`. The worker's branch dispatch was extracted into `materialize.BranchMaterializer` so the worker is a thin RabbitMQ adapter and the exact same logic backs an in-process synchronous materializer used in tests.

`SyncDraftBranchFromGit` still exists but is now only invoked from the worker path (`BranchMaterializer.MaterializeBranch`), never on the request path.

File: pkg/grpc/actions/canvases/commit_canvas_repository_files.go (hunk 7 above)
Comment (mchalapuk) [reply pkg/grpc/actions/canvases/commit_canvas_repository_files.go:174]: Done in 36b857a69. `CommitCanvasRepositoryFiles` now commits to git, registers a pending draft row (`materialize.RegisterPendingDraftVersion`), clears stale staging, and calls `RequestBranchMaterialization`. The snapshot load + node reconciliation happens in the worker; the handler returns the commit SHA optimistically.

File: pkg/grpc/actions/canvases/commit_canvas_staging.go (hunk 8 above)
Comment (mchalapuk) [reply pkg/grpc/actions/canvases/commit_canvas_staging.go]: Done in 36b857a69. `CommitCanvasStaging` now registers a pending draft row at the new commit and clears staging, then calls `RequestBranchMaterialization`; the worker materializes the snapshot. The response returns the pending version metadata optimistically (the commit SHA is authoritative).

File: pkg/grpc/actions/canvases/create_canvas.go (hunk 9 above)
Comment (mchalapuk) [reply pkg/grpc/actions/canvases/create_canvas.go]: Done in 36b857a69. `CreateCanvas` now seeds git main, creates the (intentionally empty) placeholder live row in the `pending` state, and calls `RequestBranchMaterialization`. The worker loads the snapshot, creates `workflow_nodes`/webhooks by diffing the empty placeholder against the snapshot, and promotes the live version. The response echoes the just-committed (parsed + laid-out) spec optimistically rather than reading back the empty placeholder.

File: pkg/grpc/actions/canvases/publish_canvas_version.go (hunk 11 above)
Comment (mchalapuk) [reply pkg/grpc/actions/canvases/publish_canvas_version.go]: Done in 36b857a69. `PublishCanvasVersion` (and `PublishCanvasChangeRequest`) now merge to git main, delete the draft branch, and register a pending published row at the merge commit **without** moving `LiveVersionID`. The live pointer stays put so the worker's live materialization can diff the *old* live against the snapshot to reconcile `workflow_nodes`; the worker then reuses that row (matched by commit SHA) and promotes it. Name-availability is still validated synchronously as a precondition before mutating git.

File: pkg/grpc/actions/canvases/delete_canvas_version.go (hunk 10 above)
Comment (mchalapuk) [reply pkg/grpc/actions/canvases/delete_canvas_version.go:106]: This one I intentionally left synchronous, and I think it's the odd one out.

Unlike the others, `DeleteCanvasVersion` does no heavy materialization: it deletes the git branch and then `ReconcileDraftBranchDeletionsFromGit` removes the now-orphaned draft `workflow_versions` row. That's lightweight metadata cleanup — no snapshot load, no `workflow_nodes`/webhook work — so it isn't the request-path cost the other comments were about. Keeping it synchronous also preserves read-your-writes (the draft disappears immediately on the response).

The worker still performs the same reconcile as a side effect of any branch update, so deletions converge there too. Happy to push it fully into the worker if you'd prefer strict symmetry, but I'd lean toward leaving it as-is.

**Comment by mchalapuk** [comment]
> /docs-approve

This would be a better command.

**Comment by bender-rodriguez-unit1** [comment]
Closing stale PR (older than 2 weeks). Can be reopened if needed.
